### PR TITLE
Guard the List.hd call in [Show Intros]

### PR DESCRIPTION
**Pull request by cpitclaudel**
Fix for 'Anomaly: Uncaught exception Failure("hd")' after running [Show Intros] at the end of a proof:

```
Goal True. trivial. Show Intros.
```

### CODE REVIEW

File: toplevel/vernacentries.ml [hunk 1]
```diff
@@ -102,17 +102,19 @@ let try_print_subgoals () =
...
+    let l,_= decompose_prod_assum (strip_outer_cast (pf_concl gl)) in
+    if all
+    then
+      let lid = Tactics.find_intro_names l gl in
+      msg_notice (hov 0 (prlist_with_sep  spc pr_id lid))
+    else
+      try
+        let n = List.last l in
+        msg_notice (pr_id (List.hd (Tactics.find_intro_names [n] gl)))
+      with Failure "List.last" -> ()
```
Comment (gasche) [inline toplevel/vernacentries.ml]: Is this the occasion to rewrite this dubious pattern of catching a string exception? `else if l <> []` should do the trick.

Comment (cpitclaudel) [inline toplevel/vernacentries.ml]: Sure. I've updated the PR.

**Comment by herbelin** [comment]
Ok to me, with two remarks:
- There were a policy at not using ad hoc polymorphic equality anymore. In this situation this would mean using "List.is_empty", or at worst "l != []". I would be in favor of continuing enforcing this policy whenever possible, if not too painful.
- Maybe this List.hd gls is a hint that Show Intros should apply to a goal number, though I don't know how to integrate this smoothly (thinking aloud: we have Show 2, which maybe should be 2:Show, but Show is not a tactic, so maybe a bad idea; in the case of Show Intros, this would then be Show 2 Intros, but if Show Intros is after all mainly for use from PG, maybe applicability to any goal is not necessary).

**Comment by letouzey** [comment]
Hi, I am the author of this ugly peace of code, sorry.

This command could be used by other interfaces I suppose. It allows to
insert named intros automatically (try C-c C-a C-i in pg).

Ideally I would rather have a tactical show_new_hyp_names that would apply
to any tactic and print the new hypothesis names.

And Ideall^2 I would like "as" to be reimplemented as a (predictable)
tactical renaming new hyps names only. But that would be heavily
incompatible with current behaviour (as would need less names).

P.




2015-05-04 13:17 GMT+02:00 Hugo Herbelin <notifications@github.com>:

**Comment by cpitclaudel** [comment]
Hi Hugo and Pierre,

Hugo, I'm not sure if applicability to any goal would be useful in the current state of things. I've fixed uses of polymorphic equality as directed.

Pierre, that feature in PG is really nice. I've added support in `company-coq` too, so typing `intros! RET` does the same as `C-c C-a C-i`. A general tactical might be tricky though, since sometimes the names get introduced in various branches. Perhaps something that would print a fully explicit version fo the command would be nice (as in `explicit intros` prints `intros x y H`, and `explicit induction n` prints `induction n as [ | n0 IHn]`). That sounds like a lot of work though. 

Clément.

**Comment by letouzey** [comment]
Hi Clément,
I have tried this kind of things but "as" is so adhoc that it is almost
impossible to guess the right list of hyps names even from the generated
subgoals. It has no clean semantics and applies to very few tactics.

I would need a real tactical with a clean semantics: "tac rename [intro
pattern]" would rename new hypothesis generated by tac. "New" here means
"hypothesis of which name did not appear before tac". Then it would be very
easy to generate "rename" closes automatically, even for tactics like
decompose that currently do not support "as".

P.

**Comment by cpitclaudel** [comment]
Indeed, that could be nice. It sounds like a tricky problem though.

**Pull request closed without merging**